### ISSUE TITLE
Mismatched route

### ISSUE THREAD

**Issue by jatubio** [skipperbent/simple-php-router#248]
Hello, 

I have this routes:

```
/aviso/{aviso}
/{pagina?}
/pagina/{pagina}
```

`/aviso/enviado` match `/aviso/{aviso}`
`/contacto` match `/{pagina?}`
`/pagina/usuarios` match `/pagina/{pagina}`

**But** `/avisolegal` match `/aviso/{aviso}` insteadof `/{pagina?}`

Thanks

**Comment by skipperbent** [comment]
Hello 

I'll look into this issue as soon as I get to my computer. 

Generally you should keep any routes with optional parameters at the bottom of your `routes.php` as the router will initiate the first matched routed without processing any further routes. 

Example 

`/hello/match`
`/hello/{optional?} `

/ Simon

**Comment by jatubio** [comment]
Thanks for the advice!

My routes are generated dynamically, but I'll check the order of the route `/{pagina?}`

In any case for this example, the match should not stop at `/aviso/{aviso}`. 
In fact, it loads the route like `/aviso/legal`

**Comment by skipperbent** [comment]
Hi @jatubio.

I'm trying to recreate the issue that you are experiencing. 

I've created the following test which passes fine. 

```php
TestRouter::get('/aviso/legal', 'DummyController@method1');
TestRouter::get('/aviso/{aviso}', 'DummyController@method1');
TestRouter::get('/pagina/{pagina}', 'DummyController@method1');
TestRouter::get('/{pagina?}', 'DummyController@method1');

TestRouter::debugNoReset('/aviso/optional', 'get');
$this->assertEquals('/aviso/{aviso}/', TestRouter::router()->getRequest()->getLoadedRoute()->getUrl());

TestRouter::debugNoReset('/pagina/optional', 'get');
$this->assertEquals('/pagina/{pagina}/', TestRouter::router()->getRequest()->getLoadedRoute()->getUrl());

TestRouter::debugNoReset('/optional', 'get');
$this->assertEquals('/{pagina?}/', TestRouter::router()->getRequest()->getLoadedRoute()->getUrl());

TestRouter::router()->reset();
```

Try adding this snippet in `Router.php` class line 126 in the `processRoutes` method. It will order the routes so all routes containing `?` (optional parameter) will be processed last, if no other routes have matched. Let me know if it solves your issue.

```php
usort($routes, function($a, $b) {
            if($a instanceof LoadableRoute) {
                if(strpos($a->getUrl(), '?') !== false) {
                    return 1;
                }
            }

            if($b instanceof LoadableRoute) {
                if(strpos($b->getUrl(), '?') !== false) {
                    return 1;
                }
            }

            return 1;
        });
```

Can you please send a list of all your dynamically loaded routes - including the contents of your `routes.php` file?

Thanks

/ Simon

**Comment by jatubio** [comment]
Hi @skipperbent 

Thanks for taking the time to check this issue. I'm away from my computer until next week.

How can i export one list of dynamically loaded routes?

I think, in order to test this issue, it is necessary to modify your test like this:

	// TestRouter::get('/aviso/legal', 'DummyController@method1');
	TestRouter::get('/aviso/{aviso}', 'DummyController@method1');
	TestRouter::get('/pagina/{pagina}', 'DummyController@method1');
	TestRouter::get('/{pagina?}', 'DummyController@method1');

	TestRouter::debugNoReset('/aviso/optional', 'get');
	$this->assertEquals('/aviso/{aviso}/', TestRouter::router()->getRequest()->getLoadedRoute()->getUrl());

	TestRouter::debugNoReset('/pagina/optional', 'get');
	$this->assertEquals('/pagina/{pagina}/', TestRouter::router()->getRequest()->getLoadedRoute()->getUrl());

	TestRouter::debugNoReset('/optional', 'get');
	$this->assertEquals('/{pagina?}/', TestRouter::router()->getRequest()->getLoadedRoute()->getUrl());

	// New test lines
	TestRouter::debugNoReset('/avisolegal', 'get');
	$this->assertNotEquals('/aviso/{aviso}/', TestRouter::router()->getRequest()->getLoadedRoute()->getUrl());

	TestRouter::debugNoReset('/avisolegal', 'get');
	$this->assertEquals('/{pagina?}/', TestRouter::router()->getRequest()->getLoadedRoute()->getUrl());

	TestRouter::router()->reset();

Removed first route, and added two new tests at the end

Thanks

**Comment by skipperbent** [comment]
Hi @jatubio .

The changes you made to the tests fails. The changes below passes:

```php
TestRouter::debugNoReset('/aviso/legal', 'get');
$this->assertEquals('/aviso/{aviso}/', TestRouter::router()->getRequest()->getLoadedRoute()->getUrl());

TestRouter::debugNoReset('/aviso/legal', 'get');
$this->assertNotEquals('/{pagina?}/', TestRouter::router()->getRequest()->getLoadedRoute()->getUrl());
```

`/aviso/legal` cannot match `/{pagina?}` as it contains more than one parameter. If you want it to match, you have to change the regular expression used on the route:

https://github.com/skipperbent/simple-php-router#regular-expression-route-match

Something like this:

```php
TestRouter::get('/{pagina?}', 'DummyController@method1')->setMatch('.');
```

You have to keep that route as the last route otherwise it will match all routes.

**Comment by jatubio** [comment]
Sorry Simon (@skipperbent), I think I should not have explained myself well.

I want `/avisolegal` to be equal to  `/pagina/avisolegal`.

It's not `/aviso/legal`, but `/avisolegal`, **a single parameter**.

This its why i have put that `/avisolegal` `assertNotEquals('/aviso/{aviso}/'`, but `assertEquals('/{pagina?}/'`

The problem seems to be, if you hace one single parameter what starts with the same name of a route, it will match the route.

Other Samples of incorrect matches:
`/testing` will match `/test/{anything}`
`/wordandletter` will match `/word/{word}`
etc...

Thanks

**Comment by skipperbent** [comment]
Ahh thanks for clarifying - yeah, I see what you mean! 👍 

It's definitely a bug!

Thanks - I'll try to fix this issue right away.

**Comment by jatubio** [comment]
Thank you for taking it into account!

**Comment by skipperbent** [comment]
Hi @jatubio .

The issue should be resolved in version 3.3.3, available here:
https://github.com/skipperbent/simple-php-router/releases/tag/3.3.3

Please let me know if you experiencing any further issues with the software. I really appreciate your feedback, so feel free to reopen or create a new ticket if you have any further questions.

/ Simon
